src/daemon/pcsc_context.cpp

```diff
@@ -0,0 +1,379 @@
+        {
+            // Querying buffer size and then filling buffer with two calls to SCardListReaders() is
+            // racey in pcsc-lite. Each call updates reader list from service. Will most likely fail
+            // if first call reports X readers and next call reports X+1 readers due to buffer being
+            // too small. The X+1:th reader will never be noticed. Have seen this happen so use
+            // SCARD_AUTOALLOCATE even though it requires ugly casting of buffer argument and is not
+            // exception safe (if multi_string_split() OOM:s we want to terminate anyway).
+            char *readers_multi_string;
+            auto readers_multi_string_size = static_cast<DWORD>(SCARD_AUTOALLOCATE);
+            LONG ret;
+
+            ret = SCardListReaders(context,
+                                   nullptr,
+                                   // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+                                   reinterpret_cast<LPSTR>(&readers_multi_string),
+                                   &readers_multi_string_size);
+
+            if (ret != SCARD_S_SUCCESS) {
+                if (ret != SCARD_E_NO_READERS_AVAILABLE)
+                    g_warning("Failed to list PC/SC readers: %s", pcsc_stringify_error(ret));
+
+                return {};
+            }
+
+            std::vector<std::string> readers = multi_string_split(readers_multi_string);
+
+            SCardFreeMemory(context, readers_multi_string);
+
+            return readers;
+        }
+
+        std::vector<SCARD_READERSTATE> initial_states(const std::vector<std::string> &reader_names)
+        {
+            std::vector<SCARD_READERSTATE> reader_states(1 + reader_names.size());
+
+            reader_states[NOTIFICATION_STATE_INDEX].szReader = NOTIFICATION_READER_NAME;
+            reader_states[NOTIFICATION_STATE_INDEX].dwCurrentState = SCARD_STATE_UNAWARE;
+            reader_states[NOTIFICATION_STATE_INDEX].dwEventState = SCARD_STATE_UNAWARE;
+
+            for (std::size_t i = 0; i < reader_names.size(); i++) {
+                reader_states[1 + i].szReader = reader_names[i].c_str();
+                reader_states[1 + i].dwCurrentState = SCARD_STATE_UNAWARE;
+                reader_states[1 + i].dwEventState = SCARD_STATE_UNAWARE;
+            }
+
+            return reader_states;
+        }
+
+        bool get_data_uid_supported(const SCARD_READERSTATE &state)
+        {
+            // TODO: Better parsing of ATR and find out if the "get data" command can be used to
+            //       extract UID for other cards than contactless storage cards that are identified
+            //       with this ATR. See:
+            //       - "3.1.3.2.3.2  Contactless Storage Cards" in
+            //         https://muscle.apdu.fr/www.pcscworkgroup.com/PCSC/V2/pcsc3_v2.01.09.pdf .
+            //       - https://en.wikipedia.org/wiki/Answer_to_reset
+            return state.cbAtr > 5 && state.rgbAtr[5] == 0x4f;
```

Comment:
Consider to use parentheses in expressions to specify the intent of the expression [AUTOSAR C++14, HIC++ rule 5.1.3]

Reply:
It looks to me like the AUTOSAR C++14 rule(s) are more reasonable than the HIC++ rule. AUTOSAR allows for this expression, I think, while HIC++ wants parentheses around... everything... almost.

I will rewrite this expression to something like:

```
    if (state.cbAtr > X)
        return state.rgbAtr[X] == Y;
    return false;

Reply:
But also, we have said "follow AUTOSAR C++14", not HIC++?

Reply:
Split the expression.